### PR TITLE
RT Consistency and BYO Consistency support for File and OCF sources

### PR DESCRIPTION
**Pull request by nacrooks**
This PR adds support for 

1) RT consistency and BYO consistency for OCF and file sources. File/OCF sources now follow the same logic as Kafka sources.

2) adds the option to express the consistency topic as either bytes or using an Avro schema.

3) adds support for decoding an Avro OCF consistency topic for Debezium (this is the format used by Debezium's embedded connector)

4) ties the envelope and encoding of the consistency format to the source encoding/envelope. 

5) adds testing for the new supported formats

6) removes the option to not have timestamping activated.

**Limitations**:

1) The Consistency Topic currently assumes that it will have the same format as the encoding/envelope specified in the associated SOURCE.

2) Protobuf sources expect a Source consistency topic. I will file a separate issue to handle protobufs (#2684)

3) The Timestamping code is beginning to balloon and will be refactored in a separate PR. Specifically, the code is currently single threaded. A single thread iterates over all sources to process the consistency topic associated with each of them. This is likely going to become a bottleneck.

4) The RT consistency implementation for file/Avro OCF sources requires tailing the file twice. This will be improved in a separate PR.  (#2686)

5) The testing of this PR currently emulates Debezium output files but has not been tested on an actual running Debezium instance due to a pending Debezium bug. 
(https://issues.redhat.com/browse/DBZ-1915)

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/2648)

### CODE REVIEW

**Comment by benesch** [comment]
Hey, sorry, I promise I'm going to find time to look at this today, but there is quite a bit of code and I'm struggling to follow along just given the size of the change! Is there a natural breakpoint along which it could be split into multiple PRs, or is it basically all or nothing?

**Comment by benesch** [comment]
Try as I might I'm afraid I couldn't make it through a review this large! I split out the pieces that looked ready to merge into #2699, which I think we can land immediately, and the pieces that I need to dive into more, into #2700.

File: src/coord/coord.rs [hunk 1]
```diff
@@ -454,11 +451,19 @@ where
                 }
                 Message::Worker(WorkerFeedbackWithMeta {
                     worker_id: _,
-                    message: WorkerFeedback::CreateSource(source_id, sc, consistency),
+                    message: WorkerFeedback::CreateSource(source_id, _sc),
                 }) => {
-                    ts_tx
-                        .send(TimestampMessage::Add(source_id, sc, consistency))
-                        .expect("Failed to send CREATE Instance notice to timestamper");
+                    if let Some(entry) = self.catalog.try_get_by_id(source_id.sid) {
```
Comment (nacrooks) [inline src/coord/coord.rs:456]: I had meant to ask you this @benesch: is this id guaranteed to exist? It seems to be the case, all sources are registered once in the catalog, but just want to double check

File: src/dataflow-types/types.rs [hunk 2]
```diff
@@ -458,7 +458,7 @@ impl ExternalSourceConnector {
 
 #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
 pub enum Consistency {
-    BringYourOwn(String),
+    BringYourOwn(PathBuf),
```
Comment (nacrooks) [inline src/dataflow-types/types.rs:461]: I'll revert to a string here as per your suggestion in your review. I'll make it explicit that we don't support non-utf8 paths in the conversation @benesch

**Pull request closed without merging**